### ISSUE TITLE
Using RegEx for Variants

### ISSUE THREAD

**Issue by ilocit** [WeblateOrg/weblate#4888]
We use weblate to translate amongst other things also **SAP** content.
SAP text lines can be exported from the SAP system as e.g. XLIFF.

There are special object types within an SAP system. One very common one is a so called **Data Element**.
Since the text of a Data Element can be used in different context, each **Data Element** usually has 5 texts, each with a different length restriction.

So my idea was now of course to use **Variants** to group these for translation to make them easily recognizable as belonging together.

This is what it looks like in the SAP GUI:
![image](https://user-images.githubusercontent.com/26718215/99836937-1c382700-2b67-11eb-8287-c2ecde71204b.png)

And here is the set of five strings in Weblate; but not yet grouped as variants:
![image](https://user-images.githubusercontent.com/26718215/99836685-c19ecb00-2b66-11eb-994f-4f391041e4c6.png)

Since there is a pattern that can be used, I specified this RegEx in **Settings** / **Translations** / **Variants**:
`//(SCRTEXT_S|SCRTEXT_M|SCRTEXT_L|REPTEXT|DDTEXT)`

An example of the keys of such a Data Element would be 

DTEL///ABSD/DE_INTEND_POSTBACKGR//**SCRTEXT_M** 00001
DTEL///ABSD/DE_INTEND_POSTBACKGR//**REPTEXT**   00001
DTEL///ABSD/DE_INTEND_POSTBACKGR//**SCRTEXT_L** 00001
DTEL///ABSD/DE_INTEND_POSTBACKGR//**SCRTEXT_S** 00001
DTEL///ABSD/DE_INTEND_POSTBACKGR//**DDTEXT**    00001
DTEL///ABSD/DE_INTEND_POSTBACKGR//**SCRTEXT_M** 00001

But the grouping does not work that way.
Is this because there is no "base" Key which ends with "DTEL///ABSD/DE_INTEND_POSTBACKGR"?
The "root" (unabbreviated text) would ideally be _DTEL///ABSD/DE_INTEND_POSTBACKGR_//**DDTEXT** or _DTEL///ABSD/DE_INTEND_POSTBACKGR_//**REPTEXT**

Thanks for any hints and suggestions.

**Comment by nijel** [comment]
I think it should work. The strings are grouped based on the key after the variant regex is removed.

**Comment by ilocit** [comment]
Right, it should, shouldn't it? But it doesn't....

When does this setting kick in? Only with brandnew components during the initialization, or also as kind of an Update to an existing component and/or for Components which are added via Auto Discovery after the setting was set for the "template" component? Maybe that is the problem we are facing?

Basically, can this setting be added and set to action also retroctively?

**Comment by nijel** [comment]
Whenever you change the regexp the variants mapping should be updated in the background. Maybe there is some issue with Celery and it doesn't process the queue? https://github.com/WeblateOrg/docker/issues/887 seems to indicate that...

**Comment by ilocit** [comment]
Well, I see this little icon that the Components are being updated. But once they are finished, nothing has changed. I still see all Strings seperate and not grouped:

![image](https://user-images.githubusercontent.com/26718215/99843803-696dc600-2b72-11eb-8deb-56c9379f1b12.png)

No **Variants** tab displayed @

**Comment by nijel** [comment]
I've just added tests to cover this case and it seems to work fine. Also it showed up correctly in my test project when using keys and regexp you've posted in the initial comment:

![obrazek](https://user-images.githubusercontent.com/212189/99876971-ec3a6380-2bfa-11eb-8f36-9a678b819de9.png)

**Comment by ilocit** [comment]
Thanks for doing all these tests, @nijel!

I also did some further tests. It seems that the issue is related to the number of trailing spaces which follow the expressions in the RegEx. The pattern is that `REPTEXT` is followed by 3 spaces and `DDTEXT `by 4. The others (`_S,_M,_L`) all have only one trailing space.

When I add those three or four spaces to the RegEx, weblate creates two groups of variants, when I then add a single space to each of the other three expressions, it WORKS! ;-)
(working pattern is `//(REPTEXT   |DDTEXT    |SCRTEXT_S |SCRTEXT_M |SCRTEXT_L )`)

A bit strange. I did not expect this with the RegEx, but perhaps it is my still limited knowledge of RegEx...

Thanks a mil for your help - problem solved.
Have a wonderful weekend.
